---
title: "technical-controls"
author: "A. S. Ghatpande"
date: "5/8/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Introduction

There are multiple samples (sequencing libraries) representing different types of technical controls, viz, water, DNA extraction, mock etc. The descriptions are found in the "SampleName" column of the original 'runinfo.csv" and its derivative variables 

```{r, filtering runinfo for techcontrols}
techcntrlsRuninfo <- filtered_runinfo %>% filter (grepl("DNA_extract|Mock|Water", SampleName))
```

Generating a manifest file that Qiime2 needs for importing data

```{r, generating the techcontrols manifest file}
techcntrls_manifest <- allfastqs_manifest %>% filter (sampleid %in% techcntrlsRuninfo$Run)
techcntrls_manifest %>% write_tsv("techcntrlsManifest.tsv")
```

The .tsv file needs to be a tab separated *text* file for Qiime2 to import. This was done using the foll command at commmand line:

    cp techcntrlsManifest.tsv techcntrlsManifest
  
## 7. importing the technical controls data into Qiime2

  *Note the fastq.gz files need to be in the PWD for this step*. The import was done using the foll command in the *Qiime2 conda environment*:

      qiime tools import --type 'SampleData[PairedEndSequencesWithQuality]'\
      --input-path techcntrlsManifest\
      --output-path techcntrlsdemux.qza\
      --input-format PairedEndFastqManifestPhred33V2\
      ;

The important parameters are as specified above for the data in SRA, which was already demuliplexed and sorted into per sample data (likely barcodes removed). The data was fastq format with Phred scores. We *assumed* that the Phred score is Phred33 based since Illumina moved from Phred64 a while ago (per Shilpa)

The data was successfully imported with the foll: message from Qiime2

  Imported techcntrlsManifest as PairedEndFastqManifestPhred33V2 to techcntrlsdemux.qza



## 8. summarizing and visualizing the imported demultiplexed reads data

Following along the 'moving pictures' tutorial in Qiime 2 docs, we then generate a summary of the demulitplexed data in the *Qiime2 conda environment* and the /Users/asgghar/edu/metagenomics/beiting directory:

        qiime demux summarize --i-data techcntrlsdemux.qza\
                              --o-visualization techcntrlsdemux.qzv\
                              ;
    
Qiime 2 reported: Saved Visualization to: techcntrlsdemux.qzv

  The visualization is viewed using the command below in the same directory and environment:

        qiime tools view techcntrlsdemux.qzv

  *The UPenn team used an average Phred score of 20 for each sample and removed sequences >256 nt and < 250 nt in length per the paper*



## 9. picking ASVs (amplicon sequence variants aka zero noise OTUs) using dada2 method
  DADA2 is a pipeline for detecting and correcting (correcting sequencing, PCR errors, removing PCR generated chimeric sequences where possible) Illumina amplicon sequence data. As implemented in the q2-dada2 plugin, this quality control process will additionally filter any phiX reads (commonly present in marker gene Illumina sequence data) that are identified in the sequencing data, and will filter chimeric sequences. It will also merge paired reads. For details see foll references in the ~/metagenomics subdirectory:
  
  *DADA2-paper-nature-methods2016.pdf*. Also see *comparision-of-metagenomic-methods-2020.pdf* in same directory for a comparison of 6 diff methods and this paper
  claims the DADA2 method is one of the best at what it does. 
       
       qiime dada2 denoise-paired \
        --i-demultiplexed-seqs techcntrlsdemux.qza \
        --p-trim-left-f 0 \
        --p-trim-left-r 0 \
        --p-trunc-len-f 250 \
        --p-trunc-len-r 250 \
        --o-representative-sequences techcntrls-rep-seqs.qza \
        --o-table techcntrls-table.qza \
        --o-denoising-stats techcntrls-stats.qza \
        ;
        
Qiime2 produced the output with the foll message:
Saved FeatureTable[Frequency] to: techcntrls-table.qza
Saved FeatureData[Sequence] to: techcntrls-rep-seqs.qza
Saved SampleData[DADA2Stats] to: techcntrls-stats.qza



## Including Plots

You can also embed plots, for example:

```{r pressure, echo=FALSE}
plot(pressure)
```

Note that the `echo = FALSE` parameter was added to the code chunk to prevent printing of the R code that generated the plot.
